<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Ever-heard-of-'acabion'?" data-toc-modified-id="Ever-heard-of-'acabion'?-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Ever heard of 'acabion'?</a></span></li><li><span><a href="#Brands-you've-never-heard-of-aside,-what-are-we-looking-for?" data-toc-modified-id="Brands-you've-never-heard-of-aside,-what-are-we-looking-for?-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Brands you've never heard of aside, what are we looking for?</a></span></li><li><span><a href="#There's-no-replacement-for-displacement" data-toc-modified-id="There's-no-replacement-for-displacement-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>There's no replacement for displacement</a></span></li><li><span><a href="#train_test_split" data-toc-modified-id="train_test_split-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>train_test_split</a></span></li></ul></div>

# Phase 3 Project

Our data science team has been contracted by a fairly new up and coming motorcycle manufacturer that is struggling to find their new formula for a sales success. Being a company of motorcycle riders, they have tasked our team to come up with a plan to replicate the successes of platforms like the Yamaha MT-07 and the Suzuki SV650. We will use the 'all_bikes_curated' dataset from Kaggle, curated by Emmanuel F. Werr


<img src="pictures/092220-2021-bmw-m1000rr-f.webp"  />
<center>2022 BMW M1000RR</center>

In [1]:
# Import warnings 
import warnings
warnings.filterwarnings('ignore')

# Import pandas to read our data
import pandas as pd
# 'all_bikez_curated.csv', with a 'z'
df = pd.read_csv('all_bikez_curated.csv')

# Show our data 
df.head()

Brand                Model  Year                   Category  Rating  \
0  acabion      da vinci 650-vi  2011  Prototype / concept model     3.2   
1  acabion              gtbo 55  2007                      Sport     2.6   
2  acabion  gtbo 600 daytona-vi  2011  Prototype / concept model     3.5   
3  acabion  gtbo 600 daytona-vi  2021  Prototype / concept model     NaN   
4  acabion              gtbo 70  2007  Prototype / concept model     3.1   

   Displacement (ccm)  Power (hp)  Torque (Nm) Engine cylinder Engine stroke  \
0                 NaN       804.0          NaN        Electric      Electric   
1              1300.0       541.0        420.0    In-line four   four-stroke   
2                 NaN       536.0          NaN        Electric      Electric   
3                 NaN       536.0          NaN        Electric      Electric   
4              1300.0       689.0        490.0    In-line four   four-stroke   

   ... Dry weight (kg)  Wheelbase (mm) Seat height (mm)  Front brakes  \
0  ...           420.0             NaN              NaN   Single disc   
1  ...           360.0             NaN              NaN           NaN   
2  ...           420.0             NaN              NaN   Single disc   
3  ...           420.0             NaN              NaN   Single disc   
4  ...           300.0             NaN              NaN           NaN   

   Rear brakes Front tire Rear tire Front suspension  Rear suspension  \
0  Single disc        NaN       NaN              NaN              NaN   
1          NaN        NaN       NaN              NaN              NaN   
2  Single disc        NaN       NaN              NaN              NaN   
3  Single disc        NaN       NaN              NaN              NaN   
4          NaN        NaN       NaN              NaN              NaN   

   Color options  
0            NaN  
1            NaN  
2            NaN  
3            NaN  
4   Custom made.  

[5 rows x 28 columns]

## Ever heard of 'acabion'? 

<img src="pictures/acabion_2.jpg" />
<center>Acabion GTBO450 (Not listed in dataset)</center>

Let's look at the unique values for 'Brand' as it would immediately appear that we probably aren't familiar with all of these motorcycles. Also, at ~700 hp, this thing looks like a literal DEATHTRAP.

In [2]:
# .unique() will give you a view of all of the potential outcomes in 'Brand'
df['Brand'].unique()

array(['acabion', 'access', 'ace', 'adiva', 'adler', 'adly', 'aeon',
       'aermacchi', 'agrati', 'ajp', 'ajs', 'alfer', 'alligator',
       'allstate', 'alphasports', 'alta', 'amazonas', 'american eagle',
       'american ironhorse', 'apc', 'aprilia', 'apsonic', 'arch',
       'arctic cat', 'ardie', 'ariel', 'arlen ness', 'arqin', 'askoll',
       'aspes', 'ather', 'atk', 'atlas honda', 'aurora',
       'avanturaa choppers', 'avinton', 'avon', 'azel', 'bajaj', 'balkan',
       'baltmotors', 'bamx', 'baotian', 'barossa', 'batavus', 'beeline',
       'benelli', 'bennche', 'beta', 'better', 'big bear choppers',
       'big dog', 'bimota', 'bintelli', 'black douglas', 'blackburne',
       'blata', 'bluroc', 'bmc choppers', 'bmw', 'boom trikes', 'borile',
       'boss hoss', 'bourget', 'bpg', 'brammo', 'bridgestone', 'britten',
       'brixton', 'brockhouse', 'brough superior', 'brudeli', 'bsa',
       'buccimoto', 'buell', 'bullit', 'bultaco', 'cagiva',
       'california scooter', 'can-

## Brands you've never heard of aside, what are we looking for? 

Though there have historically been a huge number of manufacturers with varying degrees of success, we're specifically looking for diamonds, regardless of country of origin, make, or model: if our goal is critical acclaim, let's then look at what the critics say, but let's first make sure that the critics actually said something. 

In [3]:
# Drop any null values in 'Rating'
df = df.dropna(subset=['Rating'])

In [4]:
# Preview first five rows and admire our work
df.head()

Brand                Model  Year                   Category  Rating  \
0   acabion      da vinci 650-vi  2011  Prototype / concept model     3.2   
1   acabion              gtbo 55  2007                      Sport     2.6   
2   acabion  gtbo 600 daytona-vi  2011  Prototype / concept model     3.5   
4   acabion              gtbo 70  2007  Prototype / concept model     3.1   
13   access      ams 8.57 lux lv  2016                        ATV     3.9   

    Displacement (ccm)  Power (hp)  Torque (Nm)  Engine cylinder  \
0                  NaN       804.0          NaN         Electric   
1               1300.0       541.0        420.0     In-line four   
2                  NaN       536.0          NaN         Electric   
4               1300.0       689.0        490.0     In-line four   
13               781.0        56.3          NaN  Single cylinder   

   Engine stroke  ... Dry weight (kg)  Wheelbase (mm) Seat height (mm)  \
0       Electric  ...           420.0             NaN              NaN   
1    four-stroke  ...           360.0             NaN              NaN   
2       Electric  ...           420.0             NaN              NaN   
4    four-stroke  ...           300.0             NaN              NaN   
13   four-stroke  ...           362.0             NaN              NaN   

    Front brakes  Rear brakes Front tire  Rear tire Front suspension  \
0    Single disc  Single disc        NaN        NaN              NaN   
1            NaN          NaN        NaN        NaN              NaN   
2    Single disc  Single disc        NaN        NaN              NaN   
4            NaN          NaN        NaN        NaN              NaN   
13   Single disc  Single disc   25/8-14   25/10-14               NaN   

    Rear suspension      Color options  
0               NaN                NaN  
1               NaN                NaN  
2               NaN                NaN  
4               NaN       Custom made.  
13              NaN  Black, white, red  

[5 rows x 28 columns]

<div class="alert alert-block alert-info">
There are a few things that we should look into prior to making a train_test_split; what do we believe will contribute to a higher rating for a motorcycle? Is there any data here that we can identify that may or may not give us those answers? Specifically we should see if we can make any sense out of all of the null values in this dataset. 
</div>

# What makes a rating, good? 

In [5]:
# .unique() will give you all of the potential scores found in the 'Rating' column 
df['Rating'].unique()

array([3.2, 2.6, 3.5, 3.1, 3.9, 2.8, 2.9, 1.9, 3.4, 2.2, 3.3, 3. , 2.7,
       2.5, 3.7, 2. , 3.6, 3.8, 2.1, 2.4, 4. , 4.1, 2.3, 1.8, 4.2, 1.7,
       4.4, 1.4, 4.3, 4.6, 1.6, 4.5, 1.5])

<div class="alert alert-block alert-info">
Since it would appear that our ratings work on a 5 point system, let's take a look at the motorcycle whose sales performance we would like to replicate: the Yamaha MT-07 and the Suzuki SV650.
</div>

# Yamaha MT-07 & Suzuki SV650

If like me, you are an avid motorcycle rider, you know, have known, or are someone on one of these bikes. 

<center>
<table><tr>
<td> <img src="pictures/yamaha mt07.jfif" alt="Drawing" style="width: 250px;"/> </td>
<td> <img src="pictures/suzuki sv650.jfif" alt="Drawing" style="width: 250px;"/> </td>
</tr></table>
MT-07 and SV650, respectively</center>
<!-- <img src="pictures/yamaha mt07.jfif" />
<img src="pictures/suzuki sv650.jfif" /> -->

In [6]:
# Find all included models that are an MT-07
df_yamaha = df.loc[df['Model'] == 'mt-07']
df_yamaha.head(1)

Brand  Model  Year    Category  Rating  Displacement (ccm)  \
35463  yamaha  mt-07  2014  Naked bike     3.5               689.0   

       Power (hp)  Torque (Nm) Engine cylinder Engine stroke  ...  \
35463        47.3         87.5            Twin   four-stroke  ...   

      Dry weight (kg)  Wheelbase (mm) Seat height (mm)  \
35463           175.0          1440.0            815.0   

                   Front brakes               Rear brakes    Front tire  \
35463  Double disc. Hydraulic.   Single disc. Hydraulic.   120/70-ZR17    

          Rear tire  Front suspension                   Rear suspension  \
35463  180/55-ZR17   Telescopic forks  Swingarm, (Link type suspension)   

       Color options  
35463   Black, white  

[1 rows x 28 columns]

In [7]:
# Find all models that have the name SV650
df_suzuki = df.loc[df['Model'] == 'sv650']
df_suzuki.head(1)

Brand  Model  Year    Category  Rating  Displacement (ccm)  \
30500  suzuki  sv650  2008  Naked bike     3.9               645.0   

       Power (hp)  Torque (Nm) Engine cylinder Engine stroke  ...  \
30500         NaN          NaN              V2   four-stroke  ...   

      Dry weight (kg)  Wheelbase (mm) Seat height (mm)  Front brakes  \
30500           168.0          1440.0            800.0   Double disc   

       Rear brakes    Front tire     Rear tire  \
30500  Single disc  120/60-ZR17   160/60-ZR17    

                                        Front suspension  \
30500  Telescopic, coil spring, oil damped, fully adj...   

                                  Rear suspension  Color options  
30500  Link-type, 7-way adjustable spring preload     Blue, Gray  

[1 rows x 28 columns]

In [8]:
# Bring in numpy
import numpy as np

In [9]:
# Average of ratings scores for the MT-07
yamaha_mean = np.mean(df_yamaha['Rating'].unique())
print('The average score of all year-model MT-07 in dataset:{}'.format(yamaha_mean))

The average score of all year-model MT-07 in dataset:3.35


In [10]:
# Average of ratings scores for the SV650
suzuki_mean = np.mean(df_suzuki['Rating'].unique())
print('The average score of all year-model SV650 in dataset:{}'.format(suzuki_mean))

The average score of all year-model SV650 in dataset:3.457142857142857


In [11]:
# Nothing fancy here, find the average of the two scores
rating_success = (yamaha_mean + suzuki_mean) / 2
print('Ultimately, this will be our ratings goal:{}'.format(rating_success))

Ultimately, this will be our ratings goal:3.4035714285714285


# Binary values for 'Rating'
Now that we have an average score for 'Rating', let's make a new column that will give us binary results, 1 being a rating above 3.4, 0 being a rating below. 

In [12]:
# Let's look for all of the unique values for the column 'Rating'
df['Rating'].unique()

array([3.2, 2.6, 3.5, 3.1, 3.9, 2.8, 2.9, 1.9, 3.4, 2.2, 3.3, 3. , 2.7,
       2.5, 3.7, 2. , 3.6, 3.8, 2.1, 2.4, 4. , 4.1, 2.3, 1.8, 4.2, 1.7,
       4.4, 1.4, 4.3, 4.6, 1.6, 4.5, 1.5])

In [13]:
# As detailed above, anything lower than 3.4 is outside of our target
rating_num = np.where(df["Rating"]>=(3.4), 1, 0)

In [14]:
# Our "Rating" column will now return a '1' for bikes that got a rating of 3.4 or above, 0 for scores
# below 3.4
rating_num

array([0, 0, 1, ..., 1, 1, 0])

In [15]:
# In order to attach this data to our current table we will need to make it into a dataframe
data = pd.DataFrame(rating_num, columns = ['Rating_'])
data.head()

Rating_
0        0
1        0
2        1
3        0
4        1

In [16]:
# Preview binary rating counts
data.value_counts()

Rating_
1          13008
0           8780
dtype: int64

In [17]:
# Drop 'Rating' column 
df = df.drop(columns = 'Rating', axis =1)
df.head()

Brand                Model  Year                   Category  \
0   acabion      da vinci 650-vi  2011  Prototype / concept model   
1   acabion              gtbo 55  2007                      Sport   
2   acabion  gtbo 600 daytona-vi  2011  Prototype / concept model   
4   acabion              gtbo 70  2007  Prototype / concept model   
13   access      ams 8.57 lux lv  2016                        ATV   

    Displacement (ccm)  Power (hp)  Torque (Nm)  Engine cylinder  \
0                  NaN       804.0          NaN         Electric   
1               1300.0       541.0        420.0     In-line four   
2                  NaN       536.0          NaN         Electric   
4               1300.0       689.0        490.0     In-line four   
13               781.0        56.3          NaN  Single cylinder   

   Engine stroke    Gearbox  ...  Dry weight (kg) Wheelbase (mm)  \
0       Electric        NaN  ...            420.0            NaN   
1    four-stroke    6-speed  ...            360.0            NaN   
2       Electric        NaN  ...            420.0            NaN   
4    four-stroke    6-speed  ...            300.0            NaN   
13   four-stroke  Automatic  ...            362.0            NaN   

    Seat height (mm) Front brakes  Rear brakes Front tire  Rear tire  \
0                NaN  Single disc  Single disc        NaN        NaN   
1                NaN          NaN          NaN        NaN        NaN   
2                NaN  Single disc  Single disc        NaN        NaN   
4                NaN          NaN          NaN        NaN        NaN   
13               NaN  Single disc  Single disc   25/8-14   25/10-14    

    Front suspension  Rear suspension      Color options  
0                NaN              NaN                NaN  
1                NaN              NaN                NaN  
2                NaN              NaN                NaN  
4                NaN              NaN       Custom made.  
13               NaN              NaN  Black, white, red  

[5 rows x 27 columns]

In [18]:
# Put our new 'Rating_' column into our dataframe
df = pd.concat([df,data], axis =1)

In [19]:
# Make sure it worked
df.head()

Brand                Model    Year                   Category  \
0  acabion      da vinci 650-vi  2011.0  Prototype / concept model   
1  acabion              gtbo 55  2007.0                      Sport   
2  acabion  gtbo 600 daytona-vi  2011.0  Prototype / concept model   
3      NaN                  NaN     NaN                        NaN   
4  acabion              gtbo 70  2007.0  Prototype / concept model   

   Displacement (ccm)  Power (hp)  Torque (Nm) Engine cylinder Engine stroke  \
0                 NaN       804.0          NaN        Electric      Electric   
1              1300.0       541.0        420.0    In-line four   four-stroke   
2                 NaN       536.0          NaN        Electric      Electric   
3                 NaN         NaN          NaN             NaN           NaN   
4              1300.0       689.0        490.0    In-line four   four-stroke   

   Gearbox  ...  Wheelbase (mm) Seat height (mm)  Front brakes  Rear brakes  \
0      NaN  ...             NaN              NaN   Single disc  Single disc   
1  6-speed  ...             NaN              NaN           NaN          NaN   
2      NaN  ...             NaN              NaN   Single disc  Single disc   
3      NaN  ...             NaN              NaN           NaN          NaN   
4  6-speed  ...             NaN              NaN           NaN          NaN   

  Front tire Rear tire Front suspension  Rear suspension  Color options  \
0        NaN       NaN              NaN              NaN            NaN   
1        NaN       NaN              NaN              NaN            NaN   
2        NaN       NaN              NaN              NaN            NaN   
3        NaN       NaN              NaN              NaN            NaN   
4        NaN       NaN              NaN              NaN   Custom made.   

   Rating_  
0      0.0  
1      0.0  
2      1.0  
3      0.0  
4      1.0  

[5 rows x 28 columns]

In [20]:
# Preview data
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 30541 entries, 0 to 38461
Data columns (total 28 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Brand                21788 non-null  object 
 1   Model                21784 non-null  object 
 2   Year                 21788 non-null  float64
 3   Category             21788 non-null  object 
 4   Displacement (ccm)   21684 non-null  float64
 5   Power (hp)           14892 non-null  float64
 6   Torque (Nm)          10064 non-null  float64
 7   Engine cylinder      21778 non-null  object 
 8   Engine stroke        21778 non-null  object 
 9   Gearbox              19569 non-null  object 
 10  Bore (mm)            17184 non-null  float64
 11  Stroke (mm)          17184 non-null  object 
 12  Fuel capacity (lts)  18986 non-null  float64
 13  Fuel system          14546 non-null  object 
 14  Fuel control         13444 non-null  object 
 15  Cooling system       19740 non-null 

# Preprocessing data for train_test_split

## There's no replacement for displacement

In [21]:
# 'Displacement (ccm)' would be how many cubic centimeters of engine displacement/size a bike's 
# engine has
print(df['Displacement (ccm)'].value_counts())
print()
print('There are {} null values in "Displacement (ccm)" '.format(df['Displacement (ccm)'].isnull().sum()))

124.0    810
125.0    727
249.0    722
49.0     644
998.0    307
        ... 
464.0      1
216.0      1
329.6      1
99.6       1
59.9       1
Name: Displacement (ccm), Length: 1083, dtype: int64

There are 8857 null values in "Displacement (ccm)" 


<div class="alert alert-block alert-info">
Even though 8857 null values in 'Displacement (ccm)' would seem like a lot, engine displacement is an incredibly subjective parameter for a motorcycle. 
</div>

In [22]:
# Drop null values for displacement
df = df.dropna(subset =['Displacement (ccm)'])

In [23]:
# Preview dataset and identify what other work needs to be done
df.head()

Brand             Model    Year                   Category  \
1   acabion           gtbo 55  2007.0                      Sport   
4   acabion           gtbo 70  2007.0  Prototype / concept model   
13   access   ams 8.57 lux lv  2016.0                        ATV   
25   access  shade xtreme 850  2018.0                        ATV   
35    adiva            ad 125  2009.0                    Scooter   

    Displacement (ccm)  Power (hp)  Torque (Nm)  Engine cylinder  \
1               1300.0       541.0        420.0     In-line four   
4               1300.0       689.0        490.0     In-line four   
13               781.0        56.3          NaN  Single cylinder   
25               781.0        57.7          NaN  Single cylinder   
35               124.0        14.8         12.0  Single cylinder   

   Engine stroke    Gearbox  ...  Wheelbase (mm) Seat height (mm)  \
1    four-stroke    6-speed  ...             NaN              NaN   
4    four-stroke    6-speed  ...             NaN              NaN   
13   four-stroke  Automatic  ...             NaN              NaN   
25   four-stroke  Automatic  ...             NaN              NaN   
35   four-stroke  Automatic  ...          1759.0              NaN   

    Front brakes  Rear brakes  Front tire   Rear tire Front suspension  \
1            NaN          NaN         NaN         NaN              NaN   
4            NaN          NaN         NaN         NaN              NaN   
13   Single disc  Single disc    25/8-14    25/10-14               NaN   
25   Single disc  Single disc    25/8-12    25/10-12       Independent   
35   Single disc  Single disc  120/70-14   140/70-14   Telescopic fork   

    Rear suspension      Color options  Rating_  
1               NaN                NaN      0.0  
4               NaN       Custom made.      1.0  
13              NaN  Black, white, red      0.0  
25      Independent       White, black      0.0  
35       Twin shock      Silver, black      1.0  

[5 rows x 28 columns]

<div class="alert alert-block alert-info">
Now that we have preprocessed one column in our dataset, let's take a look at what other columns we should consider processing. 
</div>

In [24]:
# Number of null values in df['Displacement (ccm)']
df['Displacement (ccm)'].isnull().sum()

0

In [25]:
# Preview data
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 21684 entries, 1 to 38461
Data columns (total 28 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Brand                21684 non-null  object 
 1   Model                21680 non-null  object 
 2   Year                 21684 non-null  float64
 3   Category             21684 non-null  object 
 4   Displacement (ccm)   21684 non-null  float64
 5   Power (hp)           14804 non-null  float64
 6   Torque (Nm)          10007 non-null  float64
 7   Engine cylinder      21674 non-null  object 
 8   Engine stroke        21674 non-null  object 
 9   Gearbox              19511 non-null  object 
 10  Bore (mm)            17184 non-null  float64
 11  Stroke (mm)          17184 non-null  object 
 12  Fuel capacity (lts)  18973 non-null  float64
 13  Fuel system          14544 non-null  object 
 14  Fuel control         13442 non-null  object 
 15  Cooling system       19666 non-null 

<div class="alert alert-block alert-warning">
There are a few considerations that we should make; there are 21684 total entries and we can assume that any column with fewer than that are null values. There are however a few things to keep in mind when rationalizing whether or not we should drop these values and ultimately, whether or not we should keep them at all. 
<br><br>    
Looking specifically at 'Torque (Nm)', this would imply that the motorcycle in question has at one point or another been put on a dyno to read out torque specs. While this is typically considered the "fun" value in a bike, let's first make a model with all null values in 'Torque (Nm' expunged. Once we have that model created and fitted, let's also look at a model that does not take into account torque figures. 
</div>



A dynamometer test is typically used to tell you the torque capabilities in your engine. Developed first in 1798, the technology has come a _long_ way since. <br><br>

<img src="pictures/DIY_dyno_complete1.jpg.crdownload" width = '500'/>
<center>Photo of DIY Dyno build from skrunkwerks.com</center><br><br>
Modern dynamometer testing is done by pushing air into the air filter (or turbo if you're on a drag monster) in order to simulate flow at speed and to push cool air into the radiator. By strapping the bike to the dyno, we are then able to get power readouts via sensors in the flywheel where the rear tire can put power to the device. <br><br>
I suspect that the bikes that do not have a torque reading either did not get tested due to extenuating circumstances (smaller displacement bikes, cruisers, and dirt bikes are usually not tested) or, the modern dynamometer was not available for some years of model submission. 

# Model with torque numbers intact

As suggested, we will first look at bikes that _have_ torque (nm) information. 

In [26]:
# Remove null values from df['Torque (Nm)']
df = df.dropna(subset =['Torque (Nm)'])

In [27]:
# Preview our data to see what other columns need pruning
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 10007 entries, 1 to 38298
Data columns (total 28 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Brand                10007 non-null  object 
 1   Model                10005 non-null  object 
 2   Year                 10007 non-null  float64
 3   Category             10007 non-null  object 
 4   Displacement (ccm)   10007 non-null  float64
 5   Power (hp)           9092 non-null   float64
 6   Torque (Nm)          10007 non-null  float64
 7   Engine cylinder      10001 non-null  object 
 8   Engine stroke        10001 non-null  object 
 9   Gearbox              9045 non-null   object 
 10  Bore (mm)            7764 non-null   float64
 11  Stroke (mm)          7764 non-null   object 
 12  Fuel capacity (lts)  8634 non-null   float64
 13  Fuel system          7709 non-null   object 
 14  Fuel control         5886 non-null   object 
 15  Cooling system       8778 non-null  

<div class="alert alert-block alert-warning">
There are a few considerations that we should in regards to null values and whether or not we believe these should be included in the first place. Since we have not made these values into numbers via encoding and therefore cannot yet see correlation to our target, responsibly we should only address lines with FULL information. <br><br> 
    
In other words, we should rid our dataset of null values prior to cutting out any column information. 
    
<br>
This isn't always the case, but for our purposes:
</div>

# If the value is null, it has to go

Because 'Color options' would seem to be a pretty objective thing and typically motorcycle manufacturers have up to 4-5 years to introducing new colorways per generation model of bike, this will ultimately be unnecessary for our model. 

In [28]:
# Remove df['Color options']
df = df.drop(columns = 'Color options', axis = 1)

In [29]:
# Drop null values data that we intend on using for modeling
df = df.dropna(subset =['Model', 'Power (hp)', 'Engine cylinder', 'Engine stroke', 'Gearbox', 'Bore (mm)', 'Stroke (mm)',
                       'Fuel capacity (lts)', 'Fuel system', 'Fuel control', 'Cooling system', 'Transmission type', 
                       'Dry weight (kg)', 'Wheelbase (mm)', 'Seat height (mm)', 'Front brakes', 'Rear brakes', 'Front tire',
                       'Rear tire', 'Front suspension', 'Rear suspension', 'Rating_'])

In [30]:
# Preview our new dataset
df.head()

Brand             Model    Year          Category  Displacement (ccm)  \
195  aeon          cobra 50  2012.0               ATV                49.3   
203  aeon  crossland x4 400  2012.0               ATV               346.0   
226  aeon        urban 350i  2012.0           Scooter               313.0   
361   ajp    pr4 125 enduro  2010.0  Enduro / offroad               124.0   
365   ajp        pr4 125 sm  2010.0      Super motard               124.0   

     Power (hp)  Torque (Nm)  Engine cylinder Engine stroke    Gearbox  ...  \
195         3.0          3.7  Single cylinder    two-stroke  Automatic  ...   
203        20.1         30.0  Single cylinder   four-stroke  Automatic  ...   
226        22.8         30.0  Single cylinder   four-stroke  Automatic  ...   
361        12.5          8.5  Single cylinder   four-stroke    5-speed  ...   
365        12.6          8.5  Single cylinder   four-stroke    5-speed  ...   

     Dry weight (kg) Wheelbase (mm)  Seat height (mm)  \
195            129.0         1050.0             800.0   
203            236.0         1230.0             850.0   
226            177.0         1545.0             815.0   
361            105.0         1410.0             910.0   
365            104.0         1410.0             910.0   

                       Front brakes                     Rear brakes  \
195    Expanding brake (drum brake)                     Single disc   
203                     Double disc    Expanding brake (drum brake)   
226          Single disc. Hydraulic          Single disc. Hydraulic   
361  Single disc. 2 piston calliper  Single disc. 4 piston calliper   
365   Single disc. 2 piston caliper   Single disc. 2 piston caliper   

     Front tire   Rear tire                    Front suspension  \
195     19/7-8     18/10-8   Dual hydraulic shock, Single A-arm   
203    23/7-12    23/10-12                         Double A-Arm   
226  120/70-16   140/70-15                      Telescopic fork   
361   90/90-21   120/90-18               Paioli Hydraulic fork,   
365  100/80-17   130/70-17               Paioli Hydraulic fork,   

                            Rear suspension  Rating_  
195  Single hydraulic shock, Unit swing arm      1.0  
203                               Swing Arm      1.0  
226              Dual-damper unit swing arm      1.0  
361     Sachs mono shock progressive action      1.0  
365     Sachs mono shock progressive action      1.0  

[5 rows x 27 columns]

In [31]:
# Information on our new dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1367 entries, 195 to 21511
Data columns (total 27 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Brand                1367 non-null   object 
 1   Model                1367 non-null   object 
 2   Year                 1367 non-null   float64
 3   Category             1367 non-null   object 
 4   Displacement (ccm)   1367 non-null   float64
 5   Power (hp)           1367 non-null   float64
 6   Torque (Nm)          1367 non-null   float64
 7   Engine cylinder      1367 non-null   object 
 8   Engine stroke        1367 non-null   object 
 9   Gearbox              1367 non-null   object 
 10  Bore (mm)            1367 non-null   float64
 11  Stroke (mm)          1367 non-null   object 
 12  Fuel capacity (lts)  1367 non-null   float64
 13  Fuel system          1367 non-null   object 
 14  Fuel control         1367 non-null   object 
 15  Cooling system       1367 non-null 

# Correlation to 'Rating_'

In [32]:
# What categories are highly correlated to 'Rating_'? 
preds = []
for i in df.corr()['Rating_'].index:
    if abs(df.corr()['Rating_'][i]) > 0:
        preds.append(i)

In [33]:
# Correlation to 'Rating', all null values removed
df[preds].corr()

Year  Displacement (ccm)  Power (hp)  Torque (Nm)  \
Year                 1.000000            0.101227    0.073767     0.117446   
Displacement (ccm)   0.101227            1.000000    0.671160     0.963806   
Power (hp)           0.073767            0.671160    1.000000     0.808775   
Torque (Nm)          0.117446            0.963806    0.808775     1.000000   
Bore (mm)            0.062316            0.746311    0.602769     0.747646   
Fuel capacity (lts)  0.034674            0.630263    0.543640     0.656416   
Dry weight (kg)      0.083260            0.789542    0.302619     0.702405   
Wheelbase (mm)       0.083516            0.697821    0.336337     0.642212   
Seat height (mm)    -0.099398           -0.239227    0.227368    -0.101722   
Rating_              0.035337            0.090216    0.028566     0.076461   

                     Bore (mm)  Fuel capacity (lts)  Dry weight (kg)  \
Year                  0.062316             0.034674         0.083260   
Displacement (ccm)    0.746311             0.630263         0.789542   
Power (hp)            0.602769             0.543640         0.302619   
Torque (Nm)           0.747646             0.656416         0.702405   
Bore (mm)             1.000000             0.452070         0.440687   
Fuel capacity (lts)   0.452070             1.000000         0.646671   
Dry weight (kg)       0.440687             0.646671         1.000000   
Wheelbase (mm)        0.512398             0.573633         0.786672   
Seat height (mm)      0.101386             0.016516        -0.455095   
Rating_               0.114058             0.085826         0.081160   

                     Wheelbase (mm)  Seat height (mm)   Rating_  
Year                       0.083516         -0.099398  0.035337  
Displacement (ccm)         0.697821         -0.239227  0.090216  
Power (hp)                 0.336337          0.227368  0.028566  
Torque (Nm)                0.642212         -0.101722  0.076461  
Bore (mm)                  0.512398          0.101386  0.114058  
Fuel capacity (lts)        0.573633          0.016516  0.085826  
Dry weight (kg)            0.786672         -0.455095  0.081160  
Wheelbase (mm)             1.000000         -0.296214  0.068911  
Seat height (mm)          -0.296214          1.000000 -0.071498  
Rating_                    0.068911         -0.071498  1.000000

To round out our preprocessing step, we should take out any unnecessary information, like 'Brand' and 'Model'.  

In [34]:
# Drop df['Brand']
df = df.drop(columns = 'Brand', axis =1)

In [35]:
# Drop df['Model']
df = df.drop(columns = 'Model', axis =1)

<div class="alert alert-block alert-info">
Now that we have pruned all null information from our data set let's address remaining categorical data by dummying them. 

# Dummy categorical columns

Let's split our data into a categorical variable and dummy them into binary columns.

In [36]:
# Categorical/object columns can be found from df.info(), above
categoricals = ['Category','Engine cylinder', 'Engine stroke', 'Gearbox', 'Stroke (mm)', 'Fuel system', 'Fuel control', 
                'Cooling system', 'Transmission type', 'Front brakes', 'Rear brakes', 'Front tire', 'Rear tire', 
                'Front suspension', 'Rear suspension']

# Our dataframe needs to be encoded. OHE can also be utilized, for this time we will use 
# pd.get_dummies()
df = pd.get_dummies(df, columns=categoricals)
# Preview our dummied columns
df.head()

Year  Displacement (ccm)  Power (hp)  Torque (Nm)  Bore (mm)  \
195  2012.0                49.3         3.0          3.7       40.0   
203  2012.0               346.0        20.1         30.0       82.0   
226  2012.0               313.0        22.8         30.0       78.0   
361  2010.0               124.0        12.5          8.5       56.5   
365  2010.0               124.0        12.6          8.5       59.5   

     Fuel capacity (lts)  Dry weight (kg)  Wheelbase (mm)  Seat height (mm)  \
195                  5.0            129.0          1050.0             800.0   
203                 14.0            236.0          1230.0             850.0   
226                 13.5            177.0          1545.0             815.0   
361                  7.5            105.0          1410.0             910.0   
365                  7.5            104.0          1410.0             910.0   

     Rating_  ...  \
195      1.0  ...   
203      1.0  ...   
226      1.0  ...   
361      1.0  ...   
365      1.0  ...   

     Rear suspension_Öhlins Monoshock, fully adjustable (Pre-load - External Hydraulic Device, Compression, Rebound)  \
195                                                  0                                                                 
203                                                  0                                                                 
226                                                  0                                                                 
361                                                  0                                                                 
365                                                  0                                                                 

     Rear suspension_Öhlins Racing monoshock with piggy-back, fully adjustable.  \
195                                                  0                            
203                                                  0                            
226                                                  0                            
361                                                  0                            
365                                                  0                            

     Rear suspension_Öhlins TTX monoshock with piggy-back, fully adjustable.  \
195                                                  0                         
203                                                  0                         
226                                                  0                         
361                                                  0                         
365                                                  0                         

     Rear suspension_Öhlins TTX preload-fully adjustable top and lower speed monoshock  \
195                                                  0                                   
203                                                  0                                   
226                                                  0                                   
361                                                  0                                   
365                                                  0                                   

     Rear suspension_Öhlins TTX single shock absorber with piggy-back, Electronics management system Ohlins Smart EC 2.0 (OBTi).  \
195                                                  0                                                                             
203                                                  0                                                                             
226                                                  0                                                                             
361                                                  0                                                                             
365                                                  0                      

<div class="alert alert-block alert-success">
It looks like our columns were successfully dummied. Now we will need to split our data for modeling. 
</div>

## train_test_split


In [37]:
# target is 'y'
target = df['Rating_']
# X is the new name that we will use for our data. For testing purposes we may re-use 
#'df' if it produces better results.
X = df.drop(columns = 'Rating_', axis =1)

In [38]:
#import train_test_split
from sklearn.model_selection import train_test_split

# Create variables for modeling
X_train, X_test, y_train, y_test = train_test_split (X, target, random_state = 42)

In [39]:
xx

NameError: name 'xx' is not defined

In [45]:
import pandas as pd
import numpy as np
np.random.seed(0)
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
from sklearn.model_selection import GridSearchCV
from sklearn.preprocessing import LabelEncoder
import warnings
warnings.filterwarnings('ignore')
%matplotlib inline

In [40]:
from xgboost import XGBClassifier
y_train.value_counts().sort_index()

0.0    403
1.0    622
Name: Rating_, dtype: int64

In [42]:
from sklearn.preprocessing import LabelEncoder
encoder = LabelEncoder()

# Fit and transform the training data
y_train = pd.Series(encoder.fit_transform(y_train))

# Transform the test data
y_test = pd.Series(encoder.transform(y_test))

In [46]:
# Instantiate XGBClassifier
clf = XGBClassifier()

# Fit XGBClassifier
clf.fit(X_train, y_train)

# Predict on training and test sets
training_preds = clf.predict(X_train)
test_preds = clf.predict(X_test)

# Accuracy of training and test sets
training_accuracy = accuracy_score(y_train, training_preds)
test_accuracy = accuracy_score(y_test, test_preds)

print('Training Accuracy: {:.4}%'.format(training_accuracy * 100))
print('Validation accuracy: {:.4}%'.format(test_accuracy * 100))

[14:02:42] WARNING: C:\Windows\Temp\abs_557yfx631l\croots\recipe\xgboost-split_1659548953302\work\src\learner.cc:1115: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
Training Accuracy: 94.54%
Validation accuracy: 63.45%


In [48]:
XGBClassifier().get_params().keys()

dict_keys(['objective', 'use_label_encoder', 'base_score', 'booster', 'colsample_bylevel', 'colsample_bynode', 'colsample_bytree', 'enable_categorical', 'gamma', 'gpu_id', 'importance_type', 'interaction_constraints', 'learning_rate', 'max_delta_step', 'max_depth', 'min_child_weight', 'missing', 'monotone_constraints', 'n_estimators', 'n_jobs', 'num_parallel_tree', 'predictor', 'random_state', 'reg_alpha', 'reg_lambda', 'scale_pos_weight', 'subsample', 'tree_method', 'validate_parameters', 'verbosity'])

In [51]:
param_grid = {
    'learning_rate': [0.1],
    'max_depth': [6,5],
    'min_child_weight': [1],
    'subsample': [0.7],
    'n_estimators': [100,200],
}

In [52]:
grid_clf = GridSearchCV(clf, param_grid, scoring='accuracy', cv=None, n_jobs=1)
grid_clf.fit(X_train, y_train)

best_parameters = grid_clf.best_params_

print('Grid Search found the following optimal parameters: ')
for param_name in sorted(best_parameters.keys()):
    print('%s: %r' % (param_name, best_parameters[param_name]))

training_preds = grid_clf.predict(X_train)
test_preds = grid_clf.predict(X_test)
training_accuracy = accuracy_score(y_train, training_preds)
test_accuracy = accuracy_score(y_test, test_preds)

print('')
print('Training Accuracy: {:.4}%'.format(training_accuracy * 100))
print('Validation accuracy: {:.4}%'.format(test_accuracy * 100))
# Expected Output: learning_rate: 0.1
# max_depth: 5
# min_child_weight: 1
# n_estimators: 200
# subsample: 0.7

# Training Accuracy: 91.51%
# Validation accuracy: 66.37%

[14:21:12] WARNING: C:\Windows\Temp\abs_557yfx631l\croots\recipe\xgboost-split_1659548953302\work\src\learner.cc:1115: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
[14:21:18] WARNING: C:\Windows\Temp\abs_557yfx631l\croots\recipe\xgboost-split_1659548953302\work\src\learner.cc:1115: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
[14:21:23] WARNING: C:\Windows\Temp\abs_557yfx631l\croots\recipe\xgboost-split_1659548953302\work\src\learner.cc:1115: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
[14:21:27] WARNING:

In [55]:
param_grid1 = {
    'learning_rate': [0.1],
    'max_depth': [6,5],
    'min_child_weight': [1],
    'subsample': [0.7],
    'n_estimators': [200,300,400],
}

In [56]:
grid_clf = GridSearchCV(clf, param_grid1, scoring='accuracy', cv=None, n_jobs=1)
grid_clf.fit(X_train, y_train)

best_parameters = grid_clf.best_params_

print('Grid Search found the following optimal parameters: ')
for param_name in sorted(best_parameters.keys()):
    print('%s: %r' % (param_name, best_parameters[param_name]))

training_preds = grid_clf.predict(X_train)
test_preds = grid_clf.predict(X_test)
training_accuracy = accuracy_score(y_train, training_preds)
test_accuracy = accuracy_score(y_test, test_preds)

print('')
print('Training Accuracy: {:.4}%'.format(training_accuracy * 100))
print('Validation accuracy: {:.4}%'.format(test_accuracy * 100))
# Expected Output:
# learning_rate: 0.1
# max_depth: 5
# min_child_weight: 1
# n_estimators: 300
# subsample: 0.7

# Training Accuracy: 94.05%
# Validation accuracy: 65.79%

[14:31:57] WARNING: C:\Windows\Temp\abs_557yfx631l\croots\recipe\xgboost-split_1659548953302\work\src\learner.cc:1115: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
[14:32:06] WARNING: C:\Windows\Temp\abs_557yfx631l\croots\recipe\xgboost-split_1659548953302\work\src\learner.cc:1115: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
[14:32:14] WARNING: C:\Windows\Temp\abs_557yfx631l\croots\recipe\xgboost-split_1659548953302\work\src\learner.cc:1115: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
[14:32:24] WARNING:

[14:40:08] WARNING: C:\Windows\Temp\abs_557yfx631l\croots\recipe\xgboost-split_1659548953302\work\src\learner.cc:1115: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
[14:40:36] WARNING: C:\Windows\Temp\abs_557yfx631l\croots\recipe\xgboost-split_1659548953302\work\src\learner.cc:1115: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
[14:41:04] WARNING: C:\Windows\Temp\abs_557yfx631l\croots\recipe\xgboost-split_1659548953302\work\src\learner.cc:1115: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
[14:41:34] WARNING:

In [60]:
param_grid2 = {
    'learning_rate': [0.1],
    'max_depth': [3,4,5],
    'gamma': [0,1,2],
    'min_child_weight': [1,2,3],
    'subsample': [0.7],
    'n_estimators': [200,300,400],
    
}

In [61]:
grid_clf = GridSearchCV(clf, param_grid2, scoring='accuracy', cv=None, n_jobs=1)
grid_clf.fit(X_train, y_train)

best_parameters = grid_clf.best_params_

print('Grid Search found the following optimal parameters: ')
for param_name in sorted(best_parameters.keys()):
    print('%s: %r' % (param_name, best_parameters[param_name]))

training_preds = grid_clf.predict(X_train)
test_preds = grid_clf.predict(X_test)
training_accuracy = accuracy_score(y_train, training_preds)
test_accuracy = accuracy_score(y_test, test_preds)

print('')
print('Training Accuracy: {:.4}%'.format(training_accuracy * 100))
print('Validation accuracy: {:.4}%'.format(test_accuracy * 100))
# Expected Output: learning_rate: 0.1
# max_depth: 4
# min_child_weight: 2
# n_estimators: 300
# subsample: 0.7

# Training Accuracy: 88.68%
# Validation accuracy: 67.54%

[15:20:10] WARNING: C:\Windows\Temp\abs_557yfx631l\croots\recipe\xgboost-split_1659548953302\work\src\learner.cc:1115: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
[15:20:19] WARNING: C:\Windows\Temp\abs_557yfx631l\croots\recipe\xgboost-split_1659548953302\work\src\learner.cc:1115: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
[15:20:28] WARNING: C:\Windows\Temp\abs_557yfx631l\croots\recipe\xgboost-split_1659548953302\work\src\learner.cc:1115: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
[15:20:38] WARNING:

[15:28:10] WARNING: C:\Windows\Temp\abs_557yfx631l\croots\recipe\xgboost-split_1659548953302\work\src\learner.cc:1115: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
[15:28:38] WARNING: C:\Windows\Temp\abs_557yfx631l\croots\recipe\xgboost-split_1659548953302\work\src\learner.cc:1115: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
[15:29:03] WARNING: C:\Windows\Temp\abs_557yfx631l\croots\recipe\xgboost-split_1659548953302\work\src\learner.cc:1115: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
[15:29:26] WARNING:

[15:35:35] WARNING: C:\Windows\Temp\abs_557yfx631l\croots\recipe\xgboost-split_1659548953302\work\src\learner.cc:1115: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
[15:35:52] WARNING: C:\Windows\Temp\abs_557yfx631l\croots\recipe\xgboost-split_1659548953302\work\src\learner.cc:1115: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
[15:36:09] WARNING: C:\Windows\Temp\abs_557yfx631l\croots\recipe\xgboost-split_1659548953302\work\src\learner.cc:1115: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
[15:36:26] WARNING:

[15:43:19] WARNING: C:\Windows\Temp\abs_557yfx631l\croots\recipe\xgboost-split_1659548953302\work\src\learner.cc:1115: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
[15:43:31] WARNING: C:\Windows\Temp\abs_557yfx631l\croots\recipe\xgboost-split_1659548953302\work\src\learner.cc:1115: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
[15:43:43] WARNING: C:\Windows\Temp\abs_557yfx631l\croots\recipe\xgboost-split_1659548953302\work\src\learner.cc:1115: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
[15:44:00] WARNING:

[15:51:06] WARNING: C:\Windows\Temp\abs_557yfx631l\croots\recipe\xgboost-split_1659548953302\work\src\learner.cc:1115: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
[15:51:27] WARNING: C:\Windows\Temp\abs_557yfx631l\croots\recipe\xgboost-split_1659548953302\work\src\learner.cc:1115: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
[15:51:38] WARNING: C:\Windows\Temp\abs_557yfx631l\croots\recipe\xgboost-split_1659548953302\work\src\learner.cc:1115: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
[15:51:49] WARNING:

[15:59:21] WARNING: C:\Windows\Temp\abs_557yfx631l\croots\recipe\xgboost-split_1659548953302\work\src\learner.cc:1115: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
[15:59:47] WARNING: C:\Windows\Temp\abs_557yfx631l\croots\recipe\xgboost-split_1659548953302\work\src\learner.cc:1115: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
[16:00:12] WARNING: C:\Windows\Temp\abs_557yfx631l\croots\recipe\xgboost-split_1659548953302\work\src\learner.cc:1115: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
[16:00:41] WARNING:

[16:06:09] WARNING: C:\Windows\Temp\abs_557yfx631l\croots\recipe\xgboost-split_1659548953302\work\src\learner.cc:1115: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
[16:06:22] WARNING: C:\Windows\Temp\abs_557yfx631l\croots\recipe\xgboost-split_1659548953302\work\src\learner.cc:1115: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
[16:06:36] WARNING: C:\Windows\Temp\abs_557yfx631l\croots\recipe\xgboost-split_1659548953302\work\src\learner.cc:1115: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
[16:06:50] WARNING:

[16:12:16] WARNING: C:\Windows\Temp\abs_557yfx631l\croots\recipe\xgboost-split_1659548953302\work\src\learner.cc:1115: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
[16:12:26] WARNING: C:\Windows\Temp\abs_557yfx631l\croots\recipe\xgboost-split_1659548953302\work\src\learner.cc:1115: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
[16:12:36] WARNING: C:\Windows\Temp\abs_557yfx631l\croots\recipe\xgboost-split_1659548953302\work\src\learner.cc:1115: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
[16:12:46] WARNING:

[16:21:51] WARNING: C:\Windows\Temp\abs_557yfx631l\croots\recipe\xgboost-split_1659548953302\work\src\learner.cc:1115: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
[16:22:19] WARNING: C:\Windows\Temp\abs_557yfx631l\croots\recipe\xgboost-split_1659548953302\work\src\learner.cc:1115: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
[16:22:51] WARNING: C:\Windows\Temp\abs_557yfx631l\croots\recipe\xgboost-split_1659548953302\work\src\learner.cc:1115: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
[16:23:07] WARNING:

[16:30:05] WARNING: C:\Windows\Temp\abs_557yfx631l\croots\recipe\xgboost-split_1659548953302\work\src\learner.cc:1115: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
[16:30:28] WARNING: C:\Windows\Temp\abs_557yfx631l\croots\recipe\xgboost-split_1659548953302\work\src\learner.cc:1115: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
[16:31:00] WARNING: C:\Windows\Temp\abs_557yfx631l\croots\recipe\xgboost-split_1659548953302\work\src\learner.cc:1115: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
[16:31:28] WARNING:

[16:39:37] WARNING: C:\Windows\Temp\abs_557yfx631l\croots\recipe\xgboost-split_1659548953302\work\src\learner.cc:1115: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
[16:39:58] WARNING: C:\Windows\Temp\abs_557yfx631l\croots\recipe\xgboost-split_1659548953302\work\src\learner.cc:1115: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
[16:40:20] WARNING: C:\Windows\Temp\abs_557yfx631l\croots\recipe\xgboost-split_1659548953302\work\src\learner.cc:1115: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
[16:40:42] WARNING:

[16:49:02] WARNING: C:\Windows\Temp\abs_557yfx631l\croots\recipe\xgboost-split_1659548953302\work\src\learner.cc:1115: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
[16:49:14] WARNING: C:\Windows\Temp\abs_557yfx631l\croots\recipe\xgboost-split_1659548953302\work\src\learner.cc:1115: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
[16:49:24] WARNING: C:\Windows\Temp\abs_557yfx631l\croots\recipe\xgboost-split_1659548953302\work\src\learner.cc:1115: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
[16:49:35] WARNING:

[16:55:28] WARNING: C:\Windows\Temp\abs_557yfx631l\croots\recipe\xgboost-split_1659548953302\work\src\learner.cc:1115: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
[16:55:48] WARNING: C:\Windows\Temp\abs_557yfx631l\croots\recipe\xgboost-split_1659548953302\work\src\learner.cc:1115: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
[16:56:08] WARNING: C:\Windows\Temp\abs_557yfx631l\croots\recipe\xgboost-split_1659548953302\work\src\learner.cc:1115: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
[16:56:28] WARNING:

[17:02:24] WARNING: C:\Windows\Temp\abs_557yfx631l\croots\recipe\xgboost-split_1659548953302\work\src\learner.cc:1115: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
[17:02:41] WARNING: C:\Windows\Temp\abs_557yfx631l\croots\recipe\xgboost-split_1659548953302\work\src\learner.cc:1115: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
[17:02:57] WARNING: C:\Windows\Temp\abs_557yfx631l\croots\recipe\xgboost-split_1659548953302\work\src\learner.cc:1115: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
[17:03:19] WARNING:

[17:09:53] WARNING: C:\Windows\Temp\abs_557yfx631l\croots\recipe\xgboost-split_1659548953302\work\src\learner.cc:1115: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
[17:10:06] WARNING: C:\Windows\Temp\abs_557yfx631l\croots\recipe\xgboost-split_1659548953302\work\src\learner.cc:1115: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
[17:10:24] WARNING: C:\Windows\Temp\abs_557yfx631l\croots\recipe\xgboost-split_1659548953302\work\src\learner.cc:1115: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
[17:10:42] WARNING:

[17:18:14] WARNING: C:\Windows\Temp\abs_557yfx631l\croots\recipe\xgboost-split_1659548953302\work\src\learner.cc:1115: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
[17:18:27] WARNING: C:\Windows\Temp\abs_557yfx631l\croots\recipe\xgboost-split_1659548953302\work\src\learner.cc:1115: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
[17:18:39] WARNING: C:\Windows\Temp\abs_557yfx631l\croots\recipe\xgboost-split_1659548953302\work\src\learner.cc:1115: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
[17:18:52] WARNING:

#  Baseline Model 1.0: 
First we will try a single DecisionTree

In [ ]:
#import DecisionTreeClassifier
from sklearn.tree import DecisionTreeClassifier 
dtc = DecisionTreeClassifier(random_state=42)  
# fit classifier onto training data
dtc.fit(X_train, y_train) 

In [ ]:
print('DecisionTree Model training score: {}'.format(dtc.score(X_train, y_train)))

In [ ]:
print('DecisionTree Model test score: {}'.format(dtc.score(X_test, y_test)))

<div class="alert alert-block alert-danger">
Our first model is built on a decision tree and the performance is not great. This is done with no regard to hyperparameters, so let's call this our baseline model. If that's the case, let's see how we can improve it. 
</div>

# KNN: K-Nearest-Neighbors
KNN or K-Nearest-Neighbors is a distance based estimator; we will need to scale our data prior to modeling it in order to get best results. 


In [ ]:
# This next step can be done without instantiating a pipeline but the tool makes the process simpler
from sklearn.preprocessing import StandardScaler
from sklearn.neighbors import KNeighborsClassifier
from sklearn.pipeline import Pipeline


knn_pipeline = Pipeline([('ss', StandardScaler()),
                        ('knn', KNeighborsClassifier())])

In [ ]:
# Fit our knn_pipeline onto our training data, testing data and then see scores
knn_pipeline.fit(X_train,y_train)

print('KNN Training score:{}'.format(knn_pipeline.score(X_train,y_train)))
print('KNN Test score:{}'.format(knn_pipeline.score(X_test,y_test)))

Performance on our KNN model is not bad, but let's see if there is something that can give us a better test performance. 

# Bagged Decision Tree Model: 

Bagging is a form of ensemble model training. Bagging, the term is short for bootstrap aggregating.

In [ ]:
# import BaggingClassifier
from sklearn.ensemble import BaggingClassifier
b_tree =  BaggingClassifier(DecisionTreeClassifier(criterion='entropy', max_depth=5), 
                                 n_estimators=20, random_state = 42)
# fit model on training data
b_tree.fit(X_train, y_train)

In [ ]:
# Training data score
print('Bagged model training score: {}'.format(b_tree.score(X_train, y_train)))


In [ ]:
# Testing data score
print('Bagged model test score: {}'.format(b_tree.score(X_test, y_test)))


<div class="alert alert-block alert-success">
Our model has made improvements and does not appear to be overfitting. Since these are our best scores so far we should keep them in mind but for now let's move onto models 2.0.
</div>

# DecisionTree Model 2.0: Cross-Validation

We will try using cross-validation to see if we can up the score of our base Decision Tree Model 1.0. Cross-validation assists with overfitting, and even though it did not appear that our model overfit the data, it is still worth a try.  

In [ ]:
# import cross_val_score from sklearn
from sklearn.model_selection import cross_val_score
# We will re-use 'dtc', our DecisionTreeClassifier() from earlier

dtc_cross_score = cross_val_score(dtc, X_train, y_train, cv = 5)
mean_dtc_cross_score = np.mean(dtc_cross_score)

print('Mean Cross-Validation Score:{}'.format(mean_dtc_cross_score))

<div class="alert alert-block alert-danger">
The performance on this model is worse than the first base DecisionTree that we put together. We will scrap this version for now and move forward with our Bagged Decision Tree. 
</div>

# RandomForestClassifier()
We can identify and interpret feature importance from our second ensemble model, RandomForestClassifier() 

In [ ]:
# import RandomForestClassifier
from sklearn.ensemble import RandomForestClassifier
# forest variable, random_state = 42, max_depth = 5
forest = RandomForestClassifier(random_state = 42, max_depth = 5)
# fit classifier on training data
forest.fit(X_train, y_train)

In [ ]:
print('RandomForest training score: {}'.format(forest.score(X_train, y_train)))

In [ ]:
print('RandomForest testing score: {}'.format(forest.score(X_test,y_test)))

<div class="alert alert-block alert-danger">
Though this would not appear to be a successful model, let's begin hyperparameter tuning to see if we can get a score of above, say 80% on both the training and testing data
</div>

# Hyperparameter tuning on RandomForestClassifier()

In [ ]:
RandomForestClassifier().get_params().keys()

In [ ]:
# Set a new variable for our second version of the RandomForestClassifier()
# As a reminder: 
# forest = RandomForestClassifier(random_state =42, max_depth = 5)
forest1 = RandomForestClassifier(random_state =42, max_depth =5, n_estimators=10)
    

In [ ]:
forest1.fit(X_train,y_train)
print('RandomForestClassifier training score n_estimators = 10: {}'.format(forest1.score(X_train, y_train)))
print('RandomForestClassifier test score n_estimators =10: {}'.format(forest1.score(X_test, y_test)))

In [ ]:
forest2 = RandomForestClassifier(criterion = 'gini',random_state = 42, max_depth =5, n_estimators =50)
forest2.fit(X_train,y_train)
print('RandomForestClassifier training score n_estimators = 50: {}'.format(forest2.score(X_train, y_train)))
print('RandomForestClassifier test score n_estimators =50: {}'.format(forest2.score(X_test, y_test)))

<div class="alert alert-block alert-danger">
Not exactly what we had in mind. We have created an ensemble model but are having a hard time determining parameters that will increase score. For this, let's use Random Search. 
</div>

# Random Search Training

In [ ]:
# random_grid works the same way that a param_grid would: we will see what parameters are selected as 
# the 'best' parameters and build our next model based on these results. 
random_grid = {'bootstrap': [True, False],
               'max_depth': [10,30,50,70,90,100, None],
               'max_features': ['auto', 'sqrt'], 
               'n_estimators': [200, 400, 600]    
}

In [ ]:
# This takes almost 20 minutes to run, Weird Al Yankovic's 'Sauerkraut' x 2 would be unbearable but...
from sklearn.model_selection import RandomizedSearchCV
rf = RandomForestClassifier()
rf_cv = RandomizedSearchCV(estimator =rf,
                          param_distributions = random_grid, 
                           n_iter = 100,
                          cv = 3
                          )
rf_cv.fit(X_train, y_train)
rf_cv.score(X_train, y_train)
# Expected Output: 0.697560975609756

In [ ]:
# The best parameters in this first tuned model
rf_cv.best_params_
# Expected output: {'n_estimators': 600,
#  'max_features': 'auto',
#  'max_depth': 10,
#  'bootstrap': False}

<div class="alert alert-block alert-danger">
Thinking critically on these parameters then, 600 estimators was the most in our set, let's shoot for more in the following model, max_features: 'auto', max_depth: '(something in the single digits)', and 'bootstrap': False
</div>

# Random Search Training 2.0
Again, taking into account the parameters that were selected as the 'best', let's expand on that idea by making our next set of parameters exploratory options referentially to our previous set of 'best_params'. 

In [ ]:
# Set new parameters for new random_grid1
random_grid1 = {'bootstrap': [False],
               'max_depth': [1,5,7,10],
               'max_features': ['auto'], 
               'n_estimators': [600,800,1000,1200]    
}

In [ ]:
rf_cv = RandomizedSearchCV(estimator =rf,
                          param_distributions = random_grid1, 
                           n_iter = 100,
                          cv = 3
                          )
rf_cv.fit(X_train, y_train)
rf_cv.score(X_train, y_train)
# Expected output: 0.7014634146341463

In [ ]:
rf_cv.score(X_test, y_test)
# Expected output: 0.6549707602339181

In [ ]:
rf_cv.best_params_
# Expected output: {'n_estimators': 800,
#  'max_features': 'auto',
#  'max_depth': 10,
#  'bootstrap': False}

<div class="alert alert-block alert-danger">
Well there's a slight improvement, and we have gotten closer to the best score we've gotten so far from our BaggedClassifier(DecisionTreeClassifier()) but let's keep tuning. 
</div>

# Random Search Training 3.0

In [ ]:
random_grid2 = {'bootstrap': [False],
               'max_depth': [10,11,12,13,14,15,16,17,18,19,20],
               'max_features': ['auto'], 
               'n_estimators': [700, 750, 800, 850, 900]    
}

In [ ]:
# 'Stairway to Heaven' like thrice
rf_cv = RandomizedSearchCV(estimator =rf,
                          param_distributions = random_grid2, 
                           n_iter = 100,
                          cv = 5 # Changed from 3 to 5, probably accounts for the time
                          )
rf_cv.fit(X_train, y_train)
rf_cv.score(X_train, y_train)
# Expected output: 0.7346341463414634

In [ ]:
rf_cv.score(X_test, y_test)
# Expected output: 0.652046783625731

In [ ]:
rf_cv.best_params_
#{'n_estimators': 900,
#  'max_features': 'auto',
#  'max_depth': 12,
#  'bootstrap': False}

In [ ]:
random_grid3 = {'bootstrap': [False],
               'max_depth': [12],
               'max_features': ['auto'], 
               'n_estimators': [900, 925, 950, 975]    
}

In [ ]:
# Stairway to Heaven twice?
rf_cv = RandomizedSearchCV(estimator =rf,
                          param_distributions = random_grid2, 
                           n_iter = 100,
                          cv = 3 
                          )
rf_cv.fit(X_train, y_train)
rf_cv.score(X_train, y_train)
# Expected output: 0.7404878048780488

In [ ]:
rf_cv.score(X_test, y_test)
# Expected output: 0.6549707602339181

# AdaboostClassifier
While our model is performing at a rate that could be considered passable, let's try using an AdaboostClassifier to see if we can improve a model. 

In [ ]:
# import AdaBoostClassifier from sklearn
from sklearn.ensemble import AdaBoostClassifier
AdaBoostClassifier().get_params().keys()

In [ ]:
# A pretty vanilla classifier to start
ada = AdaBoostClassifier(random_state = 42)

In [ ]:
ada.fit(X_train, y_train)
ada.score(X_train, y_train), ada.score(X_test, y_test)

In [ ]:
# ada1 is our first attempt at improving the vanilla AdaBoostClassifier()
ada1 = AdaBoostClassifier(random_state = 42, learning_rate = 1.0, n_estimators = 300 )
ada1.fit(X_train, y_train)
ada1.score(X_train, y_train), ada1.score(X_test, y_test)
# Expected output: (0.8907317073170732, 0.6432748538011696)

In [ ]:
# Trying out a learning_rate = 0.2, n_estimators = 300
ada2 = AdaBoostClassifier(random_state = 42, learning_rate = 0.1, n_estimators = 100 )
ada2.fit(X_train, y_train)
ada2.score(X_train, y_train), ada2.score(X_test, y_test)
# Expected output: (0.8126829268292682, 0.6491228070175439)

In [ ]:
# Using the proposed learning_rate = 0.000975 and n_estimators = 500
ada3 = AdaBoostClassifier(random_state = 42, learning_rate = 0.1, n_estimators = 200 )
ada3.fit(X_train, y_train)
ada3.score(X_train, y_train), ada3.score(X_test, y_test)
# Expected output: (0.6068292682926829, 0.6461988304093568)

In [ ]:
# n_estimators = 1500 to see if there is improvement by increasing n_estimators 
ada4 = AdaBoostClassifier(random_state = 42, learning_rate = 0.1, n_estimators = 150 )
ada4.fit(X_train, y_train)
ada4.score(X_train, y_train), ada4.score(X_test, y_test)
# Expected output: (0.6253658536585366, 0.6578947368421053)

<div class="alert alert-block alert-danger">
Unfortunately, since none of these models performed particularly well, we will move forward with ada4 and do some crossvalidation. First though, let's trying Gradient Boosting. 
</div>

# Gradient Boosting

In [ ]:
from sklearn.ensemble import GradientBoostingClassifier
GradientBoostingClassifier().get_params().keys()

In [ ]:
# This will serve as our vanilla GradientBoostingClassifier()
gbt = GradientBoostingClassifier(random_state =42)
gbt.fit(X_train, y_train)
gbt.score(X_train, y_train), gbt.score(X_test, y_test)

In [ ]:
y2 = y_train - gbt.predict(X_train)
y2

In [ ]:
y_preds = gbt.predict(X_train) + gbt

In [ ]:
gbt2 = GradientBoostingClassifier(random_state =42)
gbt2.fit(X_train, y2)

In [ ]:
gbt.score(X_train, y2), gbt.score(X_test, y_test)

In [ ]:
parameter_grid = {}

In [ ]:
from sklearn.model_selection import GridSearchCV
clf = GridSearchCV(GradientBoostingClassifier(max_depth = 3, random_state = 42, learning_rate =1,
                                             n_estimators = 25), parameter_grid,cv=3)

In [ ]:

gbt1 = GradientBoostingClassifier(criterion = 'friedman_mse', 
                                 learning_rate = 1, 
                                 n_estimators = 25, 
                                 random_state = 42)
gbt1.fit(X_train, y_train)
gbt1.score(X_train, y_train), gbt1.score(X_test, y_test)

In [ ]:
clf.fit(X_train, y_train)
clf.score(X_train,y_train), clf.score(X_test, y_test)

In [ ]:
clf.best_params_

In [ ]:
def my_gradient_boosting_classifier(n_estimators, X, target, random_state=42):
    # Save the iteratively trained models and residuals
    trained_weak_learners = []
    past_residuals = []
    
    # Initial conditions
    new_y = target
    past_residuals.append(new_y)
    # Iteratively train model
    for n in range(n_estimators):
        # Train the new model on past residuals (note first model trains on y)
        new_model = DecisionTreeClassifier(max_depth=2, random_state=random_state)
        new_model.fit(X,new_y)
        # Find the new residuals (used to train the next model)
        new_y = new_y - new_model.predict(X) 
        # Save the (trained) model and the new residuals
        trained_weak_learners.append(new_model)
        past_residuals.append(new_y)
        
    return trained_weak_learners, past_residuals

In [ ]:
m,ys = my_gradient_boosting_classifier(3, X_train, y_train)

In [ ]:
ys

In [ ]:
# Our mean training score
dtc_training_score = dtc_grid_search.score(X_train, y_train)

# Mean Test score
dtc_testing_score = dtc_grid_search.score(X_test, y_test)

print(f"Mean Training Score: {dtc_training_score :.2%}")
print(f"Mean Test Score: {dtc_testing_score :.2%}")
print("Best Parameter Combination Found During Grid Search:")
dtc_grid_search.best_params_

# Testing area

In [ ]:
ds = pd.read_csv('all_bikez_curated.csv')
ds.head()

In [ ]:
ds = ds.drop(columns = 'Brand', axis =1)

In [ ]:
ds = ds.drop(columns = 'Model', axis =1)

In [ ]:
ds.info()

In [ ]:
from sklearn.impute import SimpleImputer
imputer = SimpleImputer(missing_values = np.nan, strategy = 'mean')
imputer = imputer.fit(ds[['Displacement (ccm)']])
ds['Displacement (ccm)'] = imputer.transform(ds[['Displacement (ccm)']])


In [ ]:
ds['Displacement (ccm)'].isna().sum()

In [ ]:
from sklearn.pipeline import Pipeline


simple_pipeline = Pipeline([('si', SimpleImputer(missing_values = np.nan, strategy = 'mean'))])

sp = simple_pipeline

In [ ]:
sp.fit(ds[['Power (hp)']])
ds['Power (hp)'] = sp.transform(ds[['Power (hp)']])

In [ ]:
ds['Power (hp)'].isna().sum()

In [ ]:
sp.fit(ds[['Torque (Nm)']])
ds['Torque (Nm)'] = sp.transform(ds[['Torque (Nm)']])

ds['Torque (Nm)'].isna().sum()

In [ ]:
sp.fit(ds[['Bore (mm)']])
ds['Bore (mm)'] = sp.transform(ds[['Bore (mm)']])

ds['Bore (mm)'].isna().sum()

In [ ]:
sp.fit(ds[['Fuel capacity (lts)']])
ds['Fuel capacity (lts)'] = sp.transform(ds[['Fuel capacity (lts)']])

ds['Fuel capacity (lts)'].isna().sum()

In [ ]:
ds = ds.dropna(subset =['Dry weight (kg)', 'Wheelbase (mm)', 'Seat height (mm)'])

In [ ]:
ds.head()

In [ ]:
ds = ds.fillna(subset = ['Category', 'Gearbox', 'Stroke (mm)', 'Fuel system', 'Fuel control', 
                'Cooling system', 'Transmission type', 'Front brakes', 'Rear brakes', 'Front tire', 'Rear tire', 
                'Front suspension', 'Rear suspension'])

# Preprocessing step

In [ ]:
from sklearn.ensemble import RandomForestRegressor

In [ ]:
categoricals = ['Category','Engine cylinder', 'Engine stroke', 'Gearbox', 'Stroke (mm)', 'Fuel system', 'Fuel control', 
                'Cooling system', 'Transmission type', 'Front brakes', 'Rear brakes', 'Front tire', 'Rear tire', 
                'Front suspension', 'Rear suspension']

ds = pd.get_dummies(ds, columns=categoricals)

In [ ]:
ds.describe()

In [ ]:
ds = ds.drop(columns = 'Color options', axis =1)

In [ ]:
ds.head()

In [ ]:
ds.isna().sum()

In [ ]:
ds = ds.dropna(subset = ['Rating'])

In [ ]:
ds.info()

# Train_test_split

In [ ]:

y = ds['Rating']
X = ds.drop(columns = 'Rating', axis =1)

In [ ]:
len(y), len(X)

In [ ]:
ds['Rating'].unique()

In [ ]:
ds.head()

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X, y,random_state = 42)

In [ ]:
rf = RandomForestRegressor(random_state = 42)
rf.fit(X_train, y_train)

rf.score(X_train, y_train), rf.score(X_test, y_test)

In [ ]:
RandomForestRegressor().get_params().keys()

In [ ]:
# parameters = {
#     'criterion': ['squared_error', 'absolute_error', 'friedman_mse', 'poisson'],
#     'n_estimators': [100,300,500], 
#     'random_state': [42]  
# }

In [ ]:
rf1 = RandomForestRegressor(random_state = 42, n_estimators = 300)

rf1.fit(X_train, y_train)
rf1.score(X_test, y_test)

In [ ]:
rf1.score(X_train, y_train)